base/array.jl

```diff
@@ -1622,7 +1622,7 @@ function pop!(a::Vector)
     if isempty(a)
         _throw_argerror("array must be non-empty")
     end
-    item = a[end]
+    @inbounds item = last(a)
```

Comment:
This does not change the result of `@code_native`. In general, we should only add `@inbounds` if necessary to improve performance.

Reply:
do you mean the use of `last` or `@inbounds`?

Reply:
I mean `@inbounds`. `last(a)` and `a[end]` are equivalent for `Vector`s (see `@less last([1])`)

Reply:
will `unsafe_getindex` make a difference then?

Reply:
The compiler already automatically elides the bounds check. This is good. It means that, unless there's a compiler bug, this line won't segfault. Adding `@inbounds` or switching to `unsafe_` doesn't improve performance (there's already no bounds check at runtime) and so we shouldn't do it. My statement about only using `@inbounds` if necessary to improve performance actually applies to all "unsafe" constructs.

Does that make sense?

Reply:
Thanks for the elaboration. Is the behaviour documented somewhere? Why does `popat!` have `@inbounds`?
https://github.com/JuliaLang/julia/blob/2b3a0f0ab41350e420934d288b68df7b3872b782/base/array.jl#L1668-L1676

Reply:
No, which cases the compiler is able to elide bounds checks is not documented anywhere, because it is a best-effort optimization and we like to keep that sort of optimization internal (so we don't get bug reports any time the optimizations change).

`popat!` probably has `@inbounds` because the compiler doesn't or didn't used to elide that bounds check based on the `1 <= i <= length(a)` check.